Okta.Xamarin/Okta.Xamarin/OidcClient.cs

```diff
@@ -98,9 +108,140 @@ public Task<OktaStateManager> SignOutOfOktaAsync(OktaStateManager stateManager)
+        /// <returns>ClaimsPrincipal</returns>
+        public async Task<ClaimsPrincipal> GetClaimsPincipalAsync(string accessToken, string authorizationServerId = "default")
+        {
+            string userInfoJson = await GetUserInfoJsonAsync(accessToken, authorizationServerId);
+
+            UserInfo userInfo = JsonConvert.DeserializeObject<UserInfo>(userInfoJson); // OKTA-371439 added to ensure proper mapping of all claims to ClaimsPrincipal
+
+            ClaimsIdentity claimsIdentity = new ClaimsIdentity(userInfo.ToClaims(), "Okta");
+            ClaimsPrincipal claimsPrincipal = new ClaimsPrincipal(claimsIdentity);
+            return claimsPrincipal;
+        }
+
+        protected async Task<string> GetUserInfoJsonAsync(string accessToken, string authorizationServerId = "default")
+        {
+            return await PerformAuthorizationServerRequestAsync(HttpMethod.Get, "/userinfo", new Dictionary<string, string>
+            {
+                {"Authorization", $"Bearer {accessToken}" }
+            }, authorizationServerId);
+        }
+
+        protected async Task<string> PerformRequestAsync(HttpMethod httpMethod, string path, Dictionary<string, string> headers)
+        {
+            return await PerformRequestAsync(httpMethod, path, headers, new Dictionary<string, string>());
+        }
+
+        protected async Task<string> PerformRequestAsync(HttpMethod httpMethod, string path, Dictionary<string, string> headers, Dictionary<string, string> formUrlEncodedContent)
+        {
+            return await PerformRequestAsync(httpMethod, path, headers, formUrlEncodedContent.Select(kvp => kvp).ToArray());
+        }
+
+        protected virtual async Task<string> PerformRequestAsync(HttpMethod httpMethod, string path, Dictionary<string, string> headers, params KeyValuePair<string, string>[] formUrlEncodedContent)
+        {
+            FormUrlEncodedContent content = null;
+            if ((bool)formUrlEncodedContent?.Any())
+            {
+                content = new FormUrlEncodedContent(formUrlEncodedContent.ToList());
+            }
+
+            if (!path.StartsWith("/"))
+            {
+                path = $"/{path}";
+            }
+
+            var request = GetHttpRequestMessage(httpMethod, $"{GetBasePath()}{path}", headers);
+            request.Content = content;
+
+            HttpResponseMessage response = await client.SendAsync(request).ConfigureAwait(false);
+            LastApiResponse = response;
+            return await response.Content.ReadAsStringAsync().ConfigureAwait(false);
+        }
+
+        protected virtual async Task<string> PerformAuthorizationServerRequestAsync(HttpMethod httpMethod, string path, Dictionary<string, string> headers, string authorizationServerId = "default", params KeyValuePair<string, string>[] formUrlEncodedContent)
+        {
+            FormUrlEncodedContent content = null;
+            if ((bool)formUrlEncodedContent?.Any())
```

Comment:
This will throw exception when `formUrlEncodedContent` is null, it can't be casted to `bool`.

Reply:
Since `formUrlEncodedContent` is a `params` parameter it will be an empty array if no parameters are specified.  If this method is called explicitly with a `null` argument for `formUrlEncodedContent` the offending code should be changed.

Reply:
Oh, I see. I did not notice this is an param parameter. Still in this case this null check (question mark) doesn't make much sense.

Reply:
Good point.  Do you have a suggestion?

Reply:
yes, I mean it can be removed, but this is not a big deal since the parameter cannot be null